crates/matrix-sdk/src/widget_api/driver/mod.rs

```diff
@@ -0,0 +1,231 @@
+                        }
+                    }
+                }
+                None => match self.room.send_raw(req.content, &req.message_type, None).await {
+                    Ok(send_res) => Ok(SendEventResponse {
+                        room_id: self.room.room_id().to_string(),
+                        event_id: send_res.event_id.to_string(),
+                    }),
+                    Err(err) => {
+                        Err(WidgetError(format!("Could not send event with error: {}", err)))
+                    }
+                },
+            }
+        } else {
+            Err(WidgetError(format!(
+                "No capability to send event of type {} with state key {} (for room events the state key is undefined if no state key is shown the state key is \"\")",
+                req.message_type, req.state_key.unwrap_or("undefined".to_string())
+            )))
+        }
+    }
+}
+impl<W: Widget> Driver<W> {
+    fn build_event_sender(&self, filter: &Vec<Filter>) -> Option<Box<dyn handler::EventSender>> {
+        if filter.len() > 0 {
+            let s: Box<dyn handler::EventSender> =
+                Box::new(EventSender { room: self.matrix_room.clone(), filter: filter.clone() });
+            return Some(s);
+        }
+        None
+    }
+
+    fn build_event_reader(&self, filter: &Vec<Filter>) -> Option<Box<dyn handler::EventReader>> {
+        if filter.len() > 0 {
+            Some(Box::new(EventReader { room: self.matrix_room.clone(), filter: filter.clone() }))
+        } else {
+            None
+        }
+    }
+
+    fn build_event_listener(
+        &mut self,
+        filter: &Vec<Filter>,
+    ) -> Option<mpsc::UnboundedReceiver<MatrixEvent>> {
+        let (tx, rx) = mpsc::unbounded_channel::<MatrixEvent>();
+        let filter = filter.clone();
+        if filter.len() > 0 {
+            let callback = move |ev: Raw<AnySyncTimelineEvent>| {
+                let filter = filter.clone();
+                let tx = tx.clone();
+                async move {
+                    match ev.deserialize_as::<MatrixEvent>() {
+                        Ok(m_ev) => {
+                            if (&filter).clone().iter().any(|f| {
+                                f.allow_event(&m_ev.event_type, &m_ev.state_key, &m_ev.content)
+                            }) {
+                                let _= tx.send(m_ev).map_err(|err|eprintln!("Could not send sync matrix message to another thread: {err}"));
+                            }
+                        }
+                        Err(err) => {
+                            eprintln!("Could not parse AnySyncTimelineEvent as crate::widget_api::MatrixEvent: {err}");
+                        }
+                    }
+                }
+            };
+            self.add_event_handler_handle = Some(self.matrix_room.add_event_handler(callback));
+            return Some(rx);
+        }
```

Comment:
I am not sure I am seeing how. Since the function does not return Result, we could either make a very short function that with `Result` return type that needs than matching because we want to do the `eprintln`.
Or not use `?`.

Reply:
We don't need `eprintln` since it's a library code, not a binary code, so we either return an error if it's sensible and let the caller deal with it. Or, if we don't do anything with it, we silently ignore/convert it into something that is sensible.